File: app/src/main/java/com/jiangkang/ktools/web/KWebViewClient.java
package com.jiangkang.ktools.web;

import android.annotation.TargetApi;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import android.webkit.MimeTypeMap;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.jiangkang.tools.utils.DownloadUtils;
import com.jiangkang.tools.utils.FileUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;


/**
 * Created by jiangkang on 2017/9/20.
 */

public class KWebViewClient extends WebViewClient {


    private static final String TAG = "KWebViewClient";

    private WebContract.IView mContext;


    public KWebViewClient(WebContract.IView context) {
        this.mContext = context;
    }

    @Override
    public void onPageStarted(WebView view, String url, Bitmap favicon) {
        super.onPageStarted(view, url, favicon);
        view.getSettings().setBlockNetworkImage(true);
    }


    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return super.shouldOverrideUrlLoading(view, request);
    }

    @Override
    public void onPageFinished(WebView view, String url) {
        super.onPageFinished(view, url);
        view.getSettings().setBlockNetworkImage(false);
        injectJsFile(view);
    }

    private void injectJsFile(WebView view) {
//        String js = FileUtils.readFromFile("web/inject.js");
//        Log.d(TAG, "injectJsFile: \n js = " + js);
        view.loadUrl("javascript:" + "(function imgOnClick(){\n" +
                "    var imgs = document.getElementsByTagName(\"img\");\n" +
                "    for(var i = 0; i < imgs.length;i++){\n" +
                "            imgs[i].onclick = function(){\n" +
                "                jk.showBigImage(this.src);\n" +
                "            };\n" +
                "    }\n" +
                "})()");
    }


    @Override
    public void onScaleChanged(WebView view, float oldScale, float newScale) {
        super.onScaleChanged(view, oldScale, newScale);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    @Override
    public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
        //此处对文件资源，js，css等请求资源进行拦截，替换
        Log.d(TAG, "shouldInterceptRequest: request = \n" +
                "\nurl = " + request.getUrl().toString() +
                "\nmethod = " + request.getMethod() +
                "\nheaders = " + request.getRequestHeaders().toString());
//        String url = request.getUrl().toString();
//        if ((url.startsWith("https://") || url.startsWith("http://")) && (url.endsWith(".png") || url.endsWith(".jpg"))) {
//            Log.d(TAG, "拦截资源 :" + url);
//            try {
//                WebResourceResponse response = new WebResourceResponse(MimeTypeMap.getFileExtensionFromUrl(".jpg"), "utf-8", FileUtils.getInputStreamFromAssets("img/dog.jpg"));
//                return response;
//            } catch (IOException e) {
//                e.printStackTrace();
//            }
//        }
        return super.shouldInterceptRequest(view, request);
    }



}

File: hybrid/src/main/java/com/jiangkang/hybrid/web/KWebViewClient.kt
package com.jiangkang.hybrid.web

import android.annotation.TargetApi
import android.graphics.Bitmap
import android.os.Build
import android.util.Log
import android.webkit.*
import com.jiangkang.tools.utils.FileUtils


/**
 * Created by jiangkang on 2017/9/20.
 */

class KWebViewClient : WebViewClient {

    private val TAG = "KWebViewClient"

    private val mContext: WebContract.IView

    private val mWebArgs: WebArgs

    constructor(context: WebContract.IView, webArgs: WebArgs) : super() {
        mContext = context
        mWebArgs = webArgs
    }

    //TODO: Kotlin 传入的参数类型要注意为null的情况，如果不修改重载函数参数类型，会出现运行时错误
    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        super.onPageStarted(view, url, favicon)
        if (mWebArgs.isLoadImgLazy) {
            view.settings.blockNetworkImage = true
        }
    }


    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        return super.shouldOverrideUrlLoading(view, request)
    }

    override fun onPageFinished(view: WebView, url: String) {
        super.onPageFinished(view, url)
        if (mWebArgs.isLoadImgLazy) {
            view.settings.blockNetworkImage = false
        }
        injectJsFile(view)
        queryHistoryList(view)
    }

    private fun queryHistoryList(view: WebView) {
        val historyList = view.copyBackForwardList()
        if (historyList != null) {
            val size = historyList.size
            val currentIndex = historyList.currentIndex
            (0..(size - 1))
                    .map { "$it-${historyList.getItemAtIndex(it)?.title}-${historyList.getItemAtIndex(it)?.url}" }
                    .forEach {
                        Log.d(TAG, it)
                    }
        }
    }

    private fun injectJsFile(view: WebView) {
        val jsString = """
            (function imgOnClick() {
              var imgs = document.getElementsByTagName('img');
              for(var i = 0; i < imgs.length;i++){
                console.log(imgs[i].src)
                imgs[i].addEventListener("click",function(e){
                    alert(e.target.src);
                    jk.showBigImage(e.target.src);
                });
              }
            })()
            """
        view.loadUrl("javascript:$jsString")
    }

    override fun onScaleChanged(view: WebView, oldScale: Float, newScale: Float) {
        super.onScaleChanged(view, oldScale, newScale)
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
        //此处对文件资源，js，css等请求资源进行拦截，替换
        Log.d(TAG, "shouldInterceptRequest: request = \n\nurl = ${request.url}\nmethod = ${request.method}\nheaders = ${request.requestHeaders}")
        if (mWebArgs.isInterceptResources){
            val url = request.url.toString()
            if ((url.startsWith("https://") || url.startsWith("http://")) && (url.endsWith(".png") or url.endsWith(".jpg"))) {
                return WebResourceResponse(MimeTypeMap.getFileExtensionFromUrl(".jpg"), "utf-8", FileUtils.getInputStreamFromAssets("img/dog.jpg"))
            }
        }
        return super.shouldInterceptRequest(view, request)
    }

}
